File: converter.py
from dataclasses import dataclass
from pathlib import Path

from PIL import Image


@dataclass
class ConversionResult:
    succeeded: int
    failed: list[Path]


def find_webp_files(source_dir: Path) -> list[Path]:
    return sorted(
        (path for path in source_dir.iterdir() if path.is_file() and path.suffix.lower() == ".webp"),
        key=lambda path: path.name.lower(),
    )


def convert_directory(source_dir: Path, output_dir: Path) -> ConversionResult:
    succeeded = 0
    failed: list[Path] = []

    for source_path in find_webp_files(source_dir):
        try:
            with Image.open(source_path) as image:
                image.save(output_dir / f"{source_path.stem}.png", "PNG")
            succeeded += 1
        except OSError:
            failed.append(source_path)

    return ConversionResult(succeeded=succeeded, failed=failed)

File: converter.js
const fs = require('node:fs/promises');
const path = require('node:path');
const sharp = require('sharp');

async function findWebpFiles(sourceDir) {
  const files = [];

  async function visit(directory) {
    for (const entry of await fs.readdir(directory, { withFileTypes: true })) {
      const entryPath = path.join(directory, entry.name);

      if (entry.isDirectory()) {
        await visit(entryPath);
      } else if (entry.isFile() && path.extname(entry.name).toLowerCase() === '.webp') {
        files.push(entryPath);
      }
    }
  }

  await visit(sourceDir);
  return files.sort((left, right) => left.localeCompare(right));
}

async function convertDirectory(sourceDir, outputDir) {
  let succeeded = 0;
  const failed = [];

  for (const sourcePath of await findWebpFiles(sourceDir)) {
    try {
      const parsedPath = path.parse(path.relative(sourceDir, sourcePath));
      const outputPath = path.join(outputDir, parsedPath.dir, `${parsedPath.name}.png`);

      await fs.mkdir(path.dirname(outputPath), { recursive: true });
      await sharp(sourcePath).png().toFile(outputPath);
      succeeded += 1;
    } catch {
      failed.push(sourcePath);
    }
  }

  return { succeeded, failed };
}

module.exports = { convertDirectory, findWebpFiles };
